### ISSUE TITLE
In-place modification of user-provided GLM_single params

### ISSUE THREAD

**Issue by MShinkle** [cvnlab/GLMsingle#152]
**Description:**  
The `GLM_single` class is modifying the user-provided `params` dictionary in-place during the `fit` method, leading to issues if this same `params` dict is passed to a new initialization of the `GLM_single` class.

**Example code to reproduce:**

```python
import numpy as np
from glmsingle import GLM_single

n_timepoints, n_conditions, n_voxels = 100, 5, 1000
stimdur, tr = 1.0, 2.0

glm_params = {
    'wantlibrary': 1,
    'wantglmdenoise': 1,
    'wantfracridge': 1,
}

for i in range(2):
    glm = GLM_single(glm_params)

    design = [np.zeros((n_timepoints, n_conditions))]
    for t in range(n_timepoints):
        condition = np.random.choice(n_conditions)
        design[0][t, condition] = 1

    data = [np.random.rand(n_voxels, n_timepoints)]

    results = glm.fit(design, data, stimdur, tr)

    print(f"glm_params keys after fit {i+1}:")
    print(glm_params.keys())
```

This breaks on the second iteration because new keys (particularly `'firdelay'`) have been added to `glm_params` dict during the fitting.

**Proposed fix:**  
This can be fixed simply by setting `self.params` to be a copy of the user-provided `params` object, instead of the object itself.

```python
import copy

class GLM_single():
    def __init__(self, params=None):
        self.params = copy.deepcopy(params) if params is not None else {}
        # ... rest of the initialization
```

**Alternative solution:**
This could also be fixed at a user level by passing a `.copy()` of their parameters dict:

```python
for i in range(2):
    glm = GLM_single(glm_params.copy())  # Use a copy of params
    
    design = [np.zeros((n_timepoints, n_conditions))]
    for t in range(n_timepoints):
        condition = np.random.choice(n_conditions)
        design[0][t, condition] = 1

    data = [np.random.rand(n_voxels, n_timepoints)]

    results = glm.fit(design, data, stimdur, tr)

    print(f"glm_params keys after fit {i+1}:")
    print(glm_params.keys())
```

or users could reinitialize the params dict or reuse the same `GLM_single` object. However, AFAICT these solutions aren’t clearly described anywhere, and the source of this error can be hard to uncover for new users.

I can make a quick PR incorporating the deepcopy fix if that would be helpful. Thanks!

**Comment by jacob-prince** [comment]
Hi @MShinkle, thanks for uncovering this issue. It would be great if you could make a PR incorporating the deepcopy fix for the params. We can then review. Much appreciated!

**Comment by MShinkle** [comment]
Hi @kendrickkay, I'd be glad to elaborate. Specifically, the error arises from the following sequence of operations:

1. The user provides a dict of params which are used to initialize an instance of the GLM_single class.
2. The `.fit` method, when called, populates the params dict with any missing items, such as `firdelay`, `hrflibrary`, and others.
3. The user provides the same params dict during a new initialization of the GLM_single class, e.g. for a new subject.
4. This new initialization breaks due to some of the new entries added to the params dict. 

Specifically, these lines from `GLM_single.__init__`:

```python
        # Check if all opt arguments are allowed
        allowed = list(default_params.keys()) + [
            'xvalscheme',
            'sessionindicator',
            'hrflibrary',
            'hrftoassume',
            'maxpolydeg'
        ]
        for key in params.keys():
            if key not in allowed:
                raise ValueError(f"""
                Input parameter not recognized: '{key}'
                Possible input parameters are:\n{allowed}
                """)
```

raise this error:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], [line 14](vscode-notebook-cell:?execution_count=1&line=14)
      [7](vscode-notebook-cell:?execution_count=1&line=7) glm_params = {
      [8](vscode-notebook-cell:?execution_count=1&line=8)     'wantlibrary': 1,
      [9](vscode-notebook-cell:?execution_count=1&line=9)     'wantglmdenoise': 1,
     [10](vscode-notebook-cell:?execution_count=1&line=10)     'wantfracridge': 1,
     [11](vscode-notebook-cell:?execution_count=1&line=11) }
     [13](vscode-notebook-cell:?execution_count=1&line=13) for i in range(2):
---> [14](vscode-notebook-cell:?execution_count=1&line=14)     glm = GLM_single(glm_params)
     [16](vscode-notebook-cell:?execution_count=1&line=16)     design = [np.zeros((n_timepoints, n_conditions))]
     [17](vscode-notebook-cell:?execution_count=1&line=17)     for t in range(n_timepoints):

File ~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:299, in GLM_single.__init__(self, params)
    [297](https://file+.vscode-resource.vscode-cdn.net/home/matthew/Code/GLMsingle/examples/~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:297) for key in params.keys():
    [298](https://file+.vscode-resource.vscode-cdn.net/home/matthew/Code/GLMsingle/examples/~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:298)     if key not in allowed:
--> [299](https://file+.vscode-resource.vscode-cdn.net/home/matthew/Code/GLMsingle/examples/~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:299)         raise ValueError(f"""
    [300](https://file+.vscode-resource.vscode-cdn.net/home/matthew/Code/GLMsingle/examples/~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:300)         Input parameter not recognized: '{key}'
    [301](https://file+.vscode-resource.vscode-cdn.net/home/matthew/Code/GLMsingle/examples/~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:301)         Possible input parameters are:\n{allowed}
    [302](https://file+.vscode-resource.vscode-cdn.net/home/matthew/Code/GLMsingle/examples/~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:302)         """)
    [304](https://file+.vscode-resource.vscode-cdn.net/home/matthew/Code/GLMsingle/examples/~/anaconda3/envs/pomlab/lib/python3.10/site-packages/glmsingle/glmsingle.py:304) self.params = params

ValueError: 
                Input parameter not recognized: 'firdelay'
                Possible input parameters are:
['numforhrf', 'hrfthresh', 'hrffitmask', 'R2thresh', 'hrfmodel', 'n_jobs', 'n_pcs', 'n_boots', 'extra_regressors', 'wantlibrary', 'wantglmdenoise', 'wantfracridge', 'chunklen', 'wantfileoutputs', 'wantmemoryoutputs', 'wanthdf5', 'wantparametric', 'wantpercentbold', 'wantlss', 'brainthresh', 'brainR2', 'brainexclude', 'pcR2cutoff', 'pcR2cutoffmask', 'pcstop', 'fracs', 'wantautoscale', 'seed', 'suppressoutput', 'lambda', 'xvalscheme', 'sessionindicator', 'hrflibrary', 'hrftoassume', 'maxpolydeg']
```

It seems the core reason this error arises is that neither of the following keys:

- firdelay
- firpct

are in `glmsingle.defaults.default_params` or the list of additional 'allowed' keys in `GLM_single.__init__`.

## Solutions
The most straightforward fix is probably to add `firdelay` and `firpct` to either `allowed` or the `default_params`. They both have defaults already in the `fit` method:

```python
        if 'firdelay' not in params:
            params['firdelay'] = 30

        if 'firpct' not in params:
            params['firpct'] = 99
```

which could be easily added to the `default_params`. I tested this, and it resolves the error.

## On copying the params dict
I'd argue that copying the params dict is still a good idea, even if the error itself is resolved by the above fix. The full set of params can still be accessed at `glmsingle_obj.params`, while leaving the user-provided params dict untouched. This promotes a clear distinction between params the user provided vs ones which were filled in by the `.fit` method.

**Comment by iancharest** [comment]
Adding to the defaults and using copy is a good idea. Thanks for this.

**Comment by MShinkle** [comment]
Happy to help! I've applied this change the the same branch/PR (#154).

**Comment by kendrickkay** [comment]
Resolved by #154.
